### PR TITLE
Add bindings for zlib and basics of `java.util.zip`

### PR DESCRIPTION
**Pull request by Duhemm**
Support for the following classes and interfaces:

 * Adler32
 * Checksum
 * Deflater
 * DeflaterOutputStream
 * Inflater
 * InflaterInputStream

Most of the code comes from Apache Harmony.

### CODE REVIEW

**Comment by ekrich** [comment]
Nice work. Should `zlib` be in the `scalanative.runtime` package or perhaps elsewhere if you think people may want to use it directly. I'm assuming `runtime` shouldn't really be imported for user programs or libraries.

File: sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala [hunk 1]
```diff
@@ -350,6 +350,7 @@ object ScalaNativePluginInternal {
         val arch = target.split("-").head
         // we need re2 to link the re2 c wrapper (cre2.h)
         val regex = Seq("re2")
+        val zlib  = Seq("z")
```
Comment (jonas) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala]: I suggest updating the docs to mention zlib as a required library with instructions on how to install it with header files.

Comment (densh) [inline sbt-scala-native/src/main/scala/scala/scalanative/sbtplugin/ScalaNativePluginInternal.scala]: I also think that this one should be completely optional, i.e. it should be linked by default unless you zip functionality.

**Comment by densh** [comment]
@Duhemm The reason you're seeing CI failure after changing to `@link` is that the file `zlib.c` is compiled unconditionally, and that one depends on zlib. What needs to be done to make it fully optional is to conditionally exclude that file from being linked inside of `nativeLinkLL` task.

It would be really great if you can also make re2 optional in the same way.

**Comment by Duhemm** [comment]
@densh I changed `nativeLinkLL` so that `zlib.c.o` is not included if the link `z` is not present. The same for `re2`.

**Comment by densh** [comment]
@Duhemm Can you please also update setup instructions and mention that libre2 and libz are optional if you use corresponding features?

**Comment by densh** [comment]
LGTM, great work @Duhemm !

**Comment by densh** [comment]
@Duhemm Latest iteration of this PR seems to fail to link the test suite.

**Comment by Duhemm** [comment]
@densh Done!

**Pull request merged**